### PR TITLE
in-page v2 base for cleaning review

### PR DESCRIPTION
**Pull request by Benjamin-Freoua-Alma**

### CODE REVIEW

File: alma/controllers/front/payment.php [hunk 1]
```diff
@@ -28,11 +28,17 @@
...
+use Alma\PrestaShop\Utils\OrderDataTrait;
 use Alma\PrestaShop\Utils\Settings;
 
 class AlmaPaymentModuleFrontController extends ModuleFrontController
 {
+    use AjaxTrait, OrderDataTrait {
+        AjaxTrait::ajaxFail insteadof OrderDataTrait;
+    }
+
     public $ssl = true;
```
Comment (ClaireAlma) [inline alma/controllers/front/payment.php:42]: unused variable

File: alma/controllers/hook/FrontHeaderHookController.php [hunk 2]
```diff
@@ -77,17 +77,22 @@ private function injectAlmaAssets($params)
...
+        $fragmentsJsUrl = Settings::getFragmentsJsUrl();
 
         $controller = $this->context->controller;
         $content = null;
 
         if (version_compare(_PS_VERSION_, '1.7', '<')) {
-            // Cart widget
-            if (
-                Settings::showEligibilityMessage()
+            if (($controller->php_self == 'order' && $controller->step == 3) || $controller->php_self == 'order-opc') {
```
Comment (ClaireAlma) [inline alma/controllers/hook/FrontHeaderHookController.php:87]: à mettre sur plusieurs ligne pour la lisibilité

File: alma/controllers/hook/FrontHeaderHookController.php [hunk 3]
```diff
@@ -77,17 +77,22 @@ private function injectAlmaAssets($params)
         $productScriptPath = 'views/js/alma-product.js';
         $productCssPath = 'views/css/alma-product.css';
         $cartScriptPath = 'views/js/alma-cart.js';
+        $fragmentsScriptPath = 'views/js/alma-fragments.js';
+        $fragmentsJsUrl = Settings::getFragmentsJsUrl();
 
         $controller = $this->context->controller;
         $content = null;
 
         if (version_compare(_PS_VERSION_, '1.7', '<')) {
```
Comment (ClaireAlma) [inline alma/controllers/hook/FrontHeaderHookController.php:86]: la fonction est trop compliqué, il faut la diviser en plusieurs

File: alma/lib/Forms/FragmentAdminFormBuilder.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+/**
+* Class FragmentAdminFormBuilder
+*/
+class FragmentAdminFormBuilder extends AbstractAlmaAdminFormBuilder
+{
+    const ALMA_ACTIVATE_FRAGMENT = 'ALMA_ACTIVATE_FRAGMENT';
+
+    protected function configForm()
```
Comment (ClaireAlma) [inline alma/lib/Forms/FragmentAdminFormBuilder.php:38]: manques les commentaires de fonctions et de const

File: alma/lib/Model/PaymentData.php [hunk 5]
```diff
@@ -47,13 +47,13 @@ class PaymentData
     const PAYMENT_METHOD = 'alma';
 
     /**
-     * @param Cart $cart
-     * @param Context $context
-     * @param array $feePlans
-     *
+     * @param $cart
```
Comment (ClaireAlma) [inline alma/lib/Model/PaymentData.php:50]: il faut typer dans la doccomment et mettre un message d'explication des variables

File: alma/lib/Utils/Logger.php [hunk 6]
```diff
@@ -28,6 +28,8 @@
     exit;
 }
 
+use Alma\PrestaShop\API\PaymentValidationError;
+use Matrix\Exception;
```
Comment (ClaireAlma) [inline alma/lib/Utils/Logger.php:32]: wrong exception

File: alma/lib/Utils/Settings.php [hunk 7]
```diff
@@ -313,6 +315,14 @@ public static function getKeyDescriptionPaymentTrigger()
         return self::get('ALMA_DESCRIPTION_TRIGGER', PaymentOnTriggeringAdminFormBuilder::ALMA_DESCRIPTION_TRIGGER_AT_SHIPPING);
     }
 
+    /**
+     * @return bool
+     */
+    public static function activateFragment()
```
Comment (ClaireAlma) [inline alma/lib/Utils/Settings.php:321]: trop de fonctions statiques dans le code :) Il faudrait plus d'injection et d'objets

File: alma/controllers/front/payment.php (hunk 1 above)
Comment (Benjamin-Freoua-Alma) [reply alma/controllers/front/payment.php:42]: @ClaireAlma 
is used here `$this->ajaxFail();` line 76, 83, 98, 106...

File: alma/controllers/hook/FrontHeaderHookController.php (hunk 3 above)
Comment (Benjamin-Freoua-Alma) [reply alma/controllers/hook/FrontHeaderHookController.php:86]: C'est toujours le problème de refacto

File: alma/lib/Utils/Logger.php (hunk 6 above)
Comment (Benjamin-Freoua-Alma) [reply alma/lib/Utils/Logger.php:32]: Je ne sais pas comment ça a été ajouté :/

File: alma/lib/Utils/Settings.php (hunk 7 above)
Comment (Benjamin-Freoua-Alma) [reply alma/lib/Utils/Settings.php:321]: Oui c'est le problème de refacto, la je m'étais basé sur l'existant

File: alma/lib/Utils/Logger.php (hunk 6 above)
Comment (ClaireAlma) [reply alma/lib/Utils/Logger.php]: La magie de php storm :)

**Review by ClaireAlma — APPROVED** [review]

**Review by joyet-simon — APPROVED** [review]

File: alma/controllers/hook/GetContentHookController.php [hunk 8]
```diff
@@ -246,6 +247,9 @@ public function processConfiguration()
             $descriptionPaymentTrigger = Tools::getValue('ALMA_DESCRIPTION_TRIGGER');
             Settings::updateValue('ALMA_DESCRIPTION_TRIGGER', $descriptionPaymentTrigger);
 
+            $activateFragment = (bool) Tools::getValue(FragmentAdminFormBuilder::ALMA_ACTIVATE_FRAGMENT . '_ON');
+            Settings::updateValue(FragmentAdminFormBuilder::ALMA_ACTIVATE_FRAGMENT, $activateFragment);
```
Comment (Francois-Gomis) [inline alma/controllers/hook/GetContentHookController.php:251]: You use eanble for others activation  no ?

File: alma/controllers/hook/PaymentOptionsHookController.php [hunk 9]
```diff
@@ -149,8 +149,9 @@ public function run($params)
                         'desc' => $descPaymentButton,
                         'plans' => (array) $plans,
                         'deferred_trigger_limit_days' => $feePlans->$key->deferred_trigger_limit_days,
-                        'apiMode' => Settings::getActiveMode(),
+                        'apiMode' => strtoupper(Settings::getActiveMode()),
```
Comment (Francois-Gomis) [inline alma/controllers/hook/PaymentOptionsHookController.php:152]: why this change ?

File: alma/views/templates/hook/payment_button_deferred.tpl [hunk 10]
```diff
@@ -26,3 +26,6 @@
     </p>
     {include file="modules/alma/views/templates/hook/_partials/deferred.tpl" plans=$plans}
 </div>
+{if $first}
```
Comment (Francois-Gomis) [inline alma/views/templates/hook/payment_button_deferred.tpl:29]: first ?

File: alma/controllers/front/payment.php [hunk 11]
```diff
@@ -28,11 +28,17 @@
...
 use Alma\PrestaShop\Utils\Logger;
+use Alma\PrestaShop\Utils\OrderDataTrait;
 use Alma\PrestaShop\Utils\Settings;
 
 class AlmaPaymentModuleFrontController extends ModuleFrontController
 {
+    use AjaxTrait, OrderDataTrait {
+        AjaxTrait::ajaxFail insteadof OrderDataTrait;
+    }
+
```
[the remark this answers is not in the document]
Comment (Benjamin-Freoua-Alma) [reply alma/controllers/front/payment.php:41]: @Francois-Gomis Yes. So ? I need to do something ?

File: alma/controllers/front/payment.php [hunk 12]
```diff
@@ -133,6 +143,12 @@ public function postProcess()
             return;
         }
 
+        if (Settings::activateFragment()) {
+            method_exists(get_parent_class($this), 'ajaxDie')
```
[the remark this answers is not in the document]
Comment (Benjamin-Freoua-Alma) [reply alma/controllers/front/payment.php:147]: @Francois-Gomis 
I don't know, it was already like this. I suppose it was for old version of Prestashop

File: alma/views/templates/hook/payment_button_deferred.tpl (hunk 10 above)
Comment (Benjamin-Freoua-Alma) [reply alma/views/templates/hook/payment_button_deferred.tpl:29]: @Francois-Gomis
It's the first gateway of the list. I don't know why it made like this but yes, we need to improve this.
It seams to display only one block for inpage.

File: alma/controllers/hook/GetContentHookController.php [hunk 13]
```diff
@@ -246,6 +247,9 @@ public function processConfiguration()
             $descriptionPaymentTrigger = Tools::getValue('ALMA_DESCRIPTION_TRIGGER');
             Settings::updateValue('ALMA_DESCRIPTION_TRIGGER', $descriptionPaymentTrigger);
 
+            $activateFragment = (bool) Tools::getValue(FragmentAdminFormBuilder::ALMA_ACTIVATE_FRAGMENT . '_ON');
```
[the remark this answers is not in the document]
Comment (Benjamin-Freoua-Alma) [reply alma/controllers/hook/GetContentHookController.php:250]: @Francois-Gomis 
For me is a flag

File: alma/controllers/hook/GetContentHookController.php (hunk 8 above)
Comment (Benjamin-Freoua-Alma) [reply alma/controllers/hook/GetContentHookController.php:251]: @Benjamin-Freoua-Alma 
Yes for my code, but it's merge code between Arnaud and Olivier :p

File: alma/controllers/hook/PaymentOptionsHookController.php (hunk 9 above)
Comment (Benjamin-Freoua-Alma) [reply alma/controllers/hook/PaymentOptionsHookController.php:152]: @Francois-Gomis 
It's because in old version we sent the mode api in lowercase and now we need to send in uppercase for In-page v2

File: alma/controllers/front/payment.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (Francois-Gomis) [reply alma/controllers/front/payment.php:41]: I think it's not necessary to re-import him you can directly use all method and you can remove insteadof.

File: alma/controllers/front/payment.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (Francois-Gomis) [reply alma/controllers/front/payment.php:158]: `    /**
     * @deprecated deprecated since 1.7.5.0, use ajaxRender instead
     * Dies and echoes output value
     *
     * @param string|null $value
     * @param string|null $controller
     * @param string|null $method
     *
     * @throws PrestaShopException
     */
    protected function ajaxDie($value = null, $controller = null, $method = null)
    {
        $this->ajaxRender($value, $controller, $method);
        exit;
    }
`

[the remark this answers is not in the document]
Comment (Francois-Gomis) [reply alma/controllers/front/payment.php:158]: https://github.com/PrestaShop/PrestaShop/blob/develop/classes/controller/Controller.php

[the remark this answers is not in the document]
Comment (Francois-Gomis) [reply alma/controllers/front/payment.php:158]: First AjaxDie() in V1.6.0.12 : 
https://github.com/PrestaShop/PrestaShop/blob/1.6.0.12/classes/controller/Controller.php

File: alma/controllers/hook/DisplayPaymentHookController.php [hunk 14]
```diff
@@ -180,8 +180,9 @@ private function displayAlmaPaymentOption($paymentOption)
             [
                 'options' => $paymentOption,
                 'old_prestashop_version' => version_compare(_PS_VERSION_, '1.6', '<'),
-                'apiMode' => Settings::getActiveMode(),
+                'apiMode' => strtoupper(Settings::getActiveMode()),
```
[the remark this answers is not in the document]
Comment (Francois-Gomis) [reply alma/controllers/hook/DisplayPaymentHookController.php:183]: Ok

File: alma/controllers/front/payment.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (Francois-Gomis) [reply alma/controllers/front/payment.php:158]: Can you create a method with the good method with version comapre, or with method_exits with version comment ( upper than 1.7.5.0 - between 1.6.0.12 and 1.7.5.0 - before 1.6..0.12 ?

File: alma/views/templates/hook/payment_button_deferred.tpl (hunk 10 above)
Comment (Francois-Gomis) [reply alma/views/templates/hook/payment_button_deferred.tpl:29]: we can together investigate this "$first", it's probably to have only 1 Block but I think it's important to understand why and comment/rename this variable

File: alma/controllers/front/payment.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (Francois-Gomis) [reply alma/controllers/front/payment.php:41]: OK

File: alma/controllers/front/payment.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (Benjamin-Freoua-Alma) [reply alma/controllers/front/payment.php:158]: 👍

**Comment by Benjamin-Freoua-Alma** [comment]
@Francois-Gomis @ClaireAlma @joyet-simon 
Thanks for your review. I fixed all your request.

File: alma/controllers/hook/DisplayPaymentHookController.php [hunk 15]
```diff
@@ -70,7 +70,7 @@ public function run($params)
             Tools::ps_round((float) $this->context->cart->getOrderTotal(true, Cart::BOTH), 2)
         );
 
-        foreach ($installmentPlans as $plan) {
+        foreach ($installmentPlans as $keyPlan => $plan) {
             $installment = $plan->installmentsCount;
             $key = "general_{$installment}_{$plan->deferredDays}_{$plan->deferredMonths}";
```
Comment (Francois-Gomis) [inline alma/controllers/hook/DisplayPaymentHookController.php:75]: Do you have a "Utils" method to generate the key  always in the same way ?

**Review by Francois-Gomis — APPROVED** [review]

File: alma/controllers/hook/DisplayPaymentHookController.php (hunk 15 above)
Comment (Benjamin-Freoua-Alma) [reply alma/controllers/hook/DisplayPaymentHookController.php:75]: @Francois-Gomis 
I don't think, but the key is just for the id in html and js code to make an uniq identity. What do you mean by the same way ?

Comment (Francois-Gomis) [reply alma/controllers/hook/DisplayPaymentHookController.php:75]: You use the same "method" in paymentOptionsHookController .

Comment (Benjamin-Freoua-Alma) [reply alma/controllers/hook/DisplayPaymentHookController.php]: @Francois-Gomis 
I created a new method `keyForInstallmentPlan()` because `keyForFeePlan()` didn't match for this feePlan.

**Review by Francois-Gomis — APPROVED** [review]

**Review by ClaireAlma — APPROVED** [review]

**Review by joyet-simon — APPROVED** [review]

**Review by joyet-simon — APPROVED** [review]
